### ISSUE TITLE
Pedantic lints for undesired method of handling overflow

### ISSUE THREAD

**Issue by kayabaNerve** [rust-lang/rust-clippy#15766]
### What it does

`wrap_on_overflow` would raise a warning if `wrapping_add` (or similar methods) are invoked.

`saturate_on_overflow` would raise a warning if `saturating_add` (or similar methods) are invoked.

### Advantage

In my work, I regularly control the behavior on overflow, but I prefer using `saturating_*`. Unfortunately, for whatever reason, I continue to accidentally type (and thankfully seem to catch) `wrapping_*`. I'm concerned for the day I write `wrapping_` and DON'T catch it. I also have yet to want, in my repository, wrap on overflow and accordingly am fine banning it entirely (without the `allow` pragma).

I believe a pedantic lint to explicitly declare saturation vs wrap as the preferred way to handle overflow is worthwhile accordingly. This would be approximate to the existing lints establishing preferences on endianess.

### Example

```rust
<code>
```

Could be written as:

```rust
<code>
```

**Comment by camuward** [comment]
You can disallow the `wrapping_*` methods for each integer type, though it's a lot of repetition. In your `clippy.toml`:

```toml
[[disallowed-methods]]
path = "i8::wrapping_add"
reason = "..."
replacement = "saturating_add"

# or simply

disallowed-methods = [
  "i8::wrapping_add",
  "i8::wrapping_add_unsigned",
  "i8::wrapping_div",
  # --- 8< ---
  "usize::wrapping_sub_signed",
  "core::num::NonZero::wrapping_abs",
  "core::num::NonZero::wrapping_neg",
]
```

This way you'd have to `#[allow(clippy::disallowed_method)]` to clear the warning upon each use. Note that there isn't always an equivalent saturating variant for every `wrapping_*` method. AFAIK this is the best clippy can do right now; I personally disagree that this usecase warrants a specific lint, but I admittedly don't fully understand the reasoning behind your desired behavior.

**Comment by kayabaNerve** [comment]
I don't see a reason _not_ to add restriction lints even if somewhat questionable, if reason.

In my codebase, 90% of the time, I want arithmetic to saturate when I want either saturate or wrap. There's already restriction lints for practices (my practice here is to prefer saturating arithmetic) such as to warn on BE/LE encoding (if one wants to use one specific endianess 90% of the time) or to warn on naïve arithmetic (`arithmetic_side_effects`). I just don't want to warn on naïve arithmetic in favor of checked/saturating/wrapped. I want to warn on wrapped arithmetic in favor of checked/saturating.

Thank you for the workaround presented to manually implement this today :)
